core/modules/layout/layout.module

```diff
@@ -1214,83 +1214,54 @@ function layout_get_layout_by_path($path = NULL, $router_item = NULL) {
     $router_item = menu_get_item($path);
   }
 
-  $cache = &backdrop_static(__FUNCTION__);
-
-  if (empty($cache[$router_item['path']])) {
-    // If the path is a node preview path, check if a custom layout exists for
-    // nodes and use that instead.
-    // @ todo: This adds very specific code for node previews which too closely
-    // couples Node and Layout modules. This needs to be reverted in a future
-    // commit.
-    $preview_node = NULL;
-    $node_types = node_type_get_types();
-    foreach ($node_types as $type) {
-      $preview_path = 'node/preview/' . $type->type . '/%';
-      if ($router_item['path'] == $preview_path) {
-        $tempstore_id = $router_item['map'][3];
-        $preview_node = node_get_node_tempstore($tempstore_id);
-        break;
-      }
-    }
-    if ($preview_node) {
-      $layouts = layout_load_multiple_by_path('node/%', TRUE);
-    }
-    else {
-      $layouts = layout_load_multiple_by_path($router_item['path'], TRUE);
-    }
-    $selected_layout = NULL;
-    foreach ($layouts as $layout) {
-      // Contexts must be set before the layout's access may be checked.
-      $contexts = $layout->getContexts();
-      foreach ($contexts as $context) {
-        if (isset($context->position)) {
-
-          // If on a node preview page, use the preview from tempstore.
-          // @ todo: This is also specific code for node previews which needs
-          // to be reverted in a future commit.
-          if ($preview_node) {
-            $context->setData($preview_node);
+  // Use the static cache, but based on the normal path (not system path).
+  $selected_layout = &backdrop_static(__FUNCTION__ . '/' . $router_item['href']);
+  if (isset($selected_layout)) {
+    return $selected_layout;
+  }
+
+  $layouts = layout_load_multiple_by_path($router_item['path'], TRUE);
+  backdrop_alter('layout_layouts', $layouts, $router_item);
```

Comment:
Would it be reasonable here to just add the alter invocation from within `layout_load_multiple_by_path()`? That way we avoid having to invoke after *every* call to that function?

Also, I think the alter would need a more descriptive name; how about `backdrop_alter('layout_load_multiple_by_path', $layouts, $router_item)`?

Reply:
> Would it be reasonable here to just add the alter invocation from within layout_load_multiple_by_path()? That way we avoid having to invoke after every call to that function?

I discussed that possibility in my (long) comment on the issue. Two reasons not to:
1. You need data from the normal path (which you can get from the router item), but `layout_load_multiple_by_path()` only receives the system path as an argument.
2. There's a few cases where you _don't_ want to invoke after every call. The case in function `dashboard_menu()` is an example of that.

Reply:
Ah, hadnt fully read your long comment. 🙂 

I dont think that either of those two situations would be a problem. If you invoke the hook, its up to the caller to do the checks to determine *if* they want to alter. So Node module would do something like:
```
function node_layout_layouts_alter(&$layouts, $path) {
  // Check very carefully if this is a preview path
  ...
  // Check carefully if there really is a menu item available on this path
  ...
  // If above OK, do the alter, if not, do nothing. 
}
```
Any other call to `layout_load_multiple_by_path()` would be perfectly safe then.

Reply:
The first one is still a problem. The `$path` that is passed to `layout_load_multiple_by_path()` is a _system path_, like `node/%`. But an implementer may need access to the _normal path_, like `node/1`. That is not easily available inside of `layout_load_multiple_by_path()`.

Reply:
And for the second case, it could result in some very hard-to-diagnose bugs down the road. Suppose we put the invocation inside of `layout_load_multiple_by_path()`. That means that in `dashboard_menu()`, the hook would be invoked. Now, if we're passing the router item to the hooks, we need to rebuild the router item inside of `layout_load_multiple_by_path()`, when all we have is the system path (e.g., 'node/%'). There's really no good way to do that. We could call `get_menu_item()`, but if the menu happened not to be built at that invocation, that call would cause the menus to rebuild, resulting in the infinite recursion that I mentioned in the final edit to my long comment. (Said infinite recursion is only sporadic, which is why things worked on my local but failed in testing.)

Reply:
Just noticed that I skipped over your comment:

>Also, I think the alter would need a more descriptive name; how about 'backdrop_alter('layout_load_multiple_by_path', $layouts, $router_item)'? 

That would make the hook `hook_layout_load_multiple_by_path_alter()`.

So, I guess that's a matter of personal style (which might be overridden by house style, of course). My rationale was to keep them both short (but not too short), so choosing the name to specify just what is being altered: `layouts` in one, `contexts` in the other.

Because the two hooks work together, I think there's value in having them stylistically similar in name.

What about `hook_layout_layouts_load_alter()` and `hook_layout_contexts_load_alter()`? There we've introduced the action verb into the name, but they're still not too long and are similar, reinforcing the notion that they work together.